kafka/consumer.go

```diff
@@ -117,39 +118,79 @@ func (c *Consumer) Unassign() (err error) {
 	return nil
 }
 
-// Commit offsets for currently assigned partitions
-func (c *Consumer) Commit(async bool) (err error) {
-	e := C.rd_kafka_commit(c.handle.rk, nil, bool2cint(async))
-	if e != C.RD_KAFKA_RESP_ERR_NO_ERROR {
-		return NewKafkaError(e)
+// commit offsets for specified offsets.
+// If offsets is nil the currently assigned partitions' offsets are committed.
+// This is a blocking call, caller will need to wrap in go-routine to
+// get async or throw-away behaviour.
+func (c *Consumer) commit(offsets []TopicPartition) (committed_offsets []TopicPartition, err error) {
+	var rkqu *C.rd_kafka_queue_t
+
+	rkqu = C.rd_kafka_queue_new(c.handle.rk)
+	defer C.rd_kafka_queue_destroy(rkqu)
+
+	var c_offsets *C.rd_kafka_topic_partition_list_t
+	if offsets != nil {
+		c_offsets = new_c_parts_from_TopicPartitions(offsets)
+		defer C.rd_kafka_topic_partition_list_destroy(c_offsets)
 	}
-	return nil
-}
 
-// Commit offset based on the provided message.
-func (c *Consumer) CommitMessage(m *Message, async bool) (err error) {
-	c_offsets := new_c_parts_from_TopicPartitions([]TopicPartition{m.TopicPartition})
-	defer C.rd_kafka_topic_partition_list_destroy(c_offsets)
+	c_err := C.rd_kafka_commit_queue(c.handle.rk, c_offsets, rkqu, nil, nil)
+	if c_err != C.RD_KAFKA_RESP_ERR_NO_ERROR {
+		return nil, NewKafkaError(c_err)
+	}
 
-	e := C.rd_kafka_commit(c.handle.rk, c_offsets, bool2cint(async))
-	if e != C.RD_KAFKA_RESP_ERR_NO_ERROR {
-		return NewKafkaError(e)
+	rkev := C.rd_kafka_queue_poll(rkqu, C.int(-1))
+	if rkev == nil {
+		// shouldn't happen
+		return nil, NewKafkaError(C.RD_KAFKA_RESP_ERR__DESTROY)
 	}
+	defer C.rd_kafka_event_destroy(rkev)
 
-	return nil
+	if C.rd_kafka_event_type(rkev) != C.RD_KAFKA_EVENT_OFFSET_COMMIT {
```

Comment:
We create a temporary queue just for the response so it will only ever see an EVENT_OFFSET_COMMIT. This panic is to safe-guard from bugs.